pkg/api/vm/handler.go

```diff
@@ -1164,6 +1181,150 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
 	})
 }
 
+func (h *vmActionHandler) getVmNetwork(name, vmNamespaceFallback string) (*cniv1.NetworkAttachmentDefinition, error) {
+	nadNamespacedName := kubevirtmultus.NetAttachDefNamespacedName(vmNamespaceFallback, name)
+	nad, err := h.nadCache.Get(nadNamespacedName.Namespace, nadNamespacedName.Name)
+	if err != nil {
+		return nil, err
+	}
+	return nad, nil
+}
+
+func isVmNetworkHotpluggable(nad *cniv1.NetworkAttachmentDefinition) (bool, error) {
+	if nad.DeletionTimestamp != nil {
+		return false, nil
+	}
+
+	// nads created by system are not hotpluggable
+	// they were used for special cases, like migration network or storage network
+	if util.IsNadCreatedBySystem(nad) {
+		return false, nil
+	}
+
+	conf, err := util.DecodeNadConfigToNetConf(nad)
+	if err != nil {
+		return false, err
+	}
+
+	if !conf.IsBridgeCNI() {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// addNic add a hotplug NIC with given interface name and network name.
+func (h *vmActionHandler) addNic(ctx context.Context, namespace, name string, input AddNicInput) error {
+	vm, err := h.vmCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	nad, err := h.getVmNetwork(input.NetworkName, namespace)
+	if err != nil {
+		return err
+	}
+
+	ok, err := isVmNetworkHotpluggable(nad)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("network %s is not hot-pluggable", input.NetworkName)
+	}
+
+	vmCopy := vm.DeepCopy()
+
+	newIface := kubevirtv1.Interface{
+		Name:                   input.InterfaceName,
+		MacAddress:             input.MacAddress,
+		Model:                  "virtio",
+		InterfaceBindingMethod: kubevirtv1.InterfaceBindingMethod{Bridge: &kubevirtv1.InterfaceBridge{}},
+	}
+	vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces = append(vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces, newIface)
+
+	newNetwork := kubevirtv1.Network{
+		Name:          input.InterfaceName,
+		NetworkSource: kubevirtv1.NetworkSource{Multus: &kubevirtv1.MultusNetwork{NetworkName: input.NetworkName}},
+	}
+	vmCopy.Spec.Template.Spec.Networks = append(vmCopy.Spec.Template.Spec.Networks, newNetwork)
+
+	_, err = h.vms.Update(vmCopy)
+	if err != nil {
+		return err
+	}
+
```

Comment:
add log/event to indicate the hotplug action; similar to remove

also consider to add an annotation to the vm to record the last hotplug & hot-unplug operations (two annotations), when, which nic, it will be helpful for troubleshooting

Reply:
Thanks, I think adding some logs some be enough for now. Addressed in https://github.com/harvester/harvester/pull/9395/commits/dc226c5705e7dfcce175c10b6bfb78b6a3f36ed3